VoodooI2CHID/VoodooI2CKeyboardHIDEventDriver.cpp

```diff
@@ -0,0 +1,482 @@
+                    case kHIDUsage_AppleVendorKeyboard_CapsLockDelayEnable:
+                    case kHIDUsage_AppleVendorKeyboard_PowerState:
+                    case kHIDUsage_AppleVendorKeyboard_Expose_All:
+                    case kHIDUsage_AppleVendorKeyboard_Expose_Desktop:
+                    case kHIDUsage_AppleVendorKeyboard_Brightness_Up:
+                    case kHIDUsage_AppleVendorKeyboard_Brightness_Down:
+                    case kHIDUsage_AppleVendorKeyboard_Language:
+                        store = true;
+                        break;
+                }
+            }
+            break;
+    }
+    
+    if(store) {
+        keyboard.elements->setObject(element);
+    }
+}
+
+//Keyboard : Loop through all createMatchingElements() and parse KeyboardElements
+IOReturn VoodooI2CKeyboardHIDEventDriver::parseElements() {
+    OSArray *elementArray = hid_interface->createMatchingElements();
+    keyboard.appleVendorSupported = getProperty(kIOHIDAppleVendorSupported, gIOServicePlane);
+    
+    if (!elementArray) {
+        return kIOReturnSuccess;
+    }
+
+    int maxElement = elementArray->getCount();
+    for (int i = 0; i < maxElement; i++) {
+        IOHIDElement* element   = nullptr;
+        
+        element = OSDynamicCast(IOHIDElement, elementArray->getObject(i));
+        if (!element)
+            continue;
+        
+        if (element->getType() == kIOHIDElementTypeCollection)
+            continue;
+        
+        if (element->getUsage() == 0)
+            continue;
+        
+        parseKeyboardElement(element);
+    }
+    
+    OSSafeReleaseNULL(elementArray);
+    return kIOReturnSuccess;
+}
+
+void VoodooI2CKeyboardHIDEventDriver::setKeyboardProperties()
+{
+    OSDictionary *properties = OSDictionary::withCapacity(4);
+    
+    if (!properties)
+        return;
+    
+    if (!keyboard.elements)
+        return;
+    
+    properties->setObject(kIOHIDElementKey, keyboard.elements);
+    
+    setProperty("Keyboard", properties);
+    
+exit:
+    OSSafeReleaseNULL(properties);
+}
+
+IOReturn VoodooI2CKeyboardHIDEventDriver::setPowerState(unsigned long whichState, IOService *whatDevice) {
+    return kIOPMAckImplied;
+}
+
+bool VoodooI2CKeyboardHIDEventDriver::start(IOService* provider) {
+    if (!super::start(provider))
+        return false;
+    
+    work_loop = getWorkLoop();
+    
+    if (!work_loop)
+        return false;
+    
+    work_loop->retain();
+    
+    command_gate = IOCommandGate::commandGate(this);
+    if (!command_gate) {
+        return false;
+    }
+    
+    work_loop->addEventSource(command_gate);
+    setProperty("VoodooI2CServices Supported", kOSBooleanTrue);
+
+    return true;
+}
+
+IOReturn VoodooI2CKeyboardHIDEventDriver::message(UInt32 type, IOService* provider, void *argument) {
+    switch (type) {
+        case kKeyboardKeyPressTime:
+        {
+            //  Remember last time key was pressed
+            key_time = *((uint64_t*)argument);
+#if DEBUG
```

Comment:
Can you please remove this? I am pretty sure `DEBUG` is not used anywhere else in VoodooI2C.

This is also essentially a keylogger if the print statement is left here.